megabus/src/main/java/com/bazaarvoice/megabus/DefaultMegabusSource.java

```diff
@@ -0,0 +1,85 @@
+package com.bazaarvoice.megabus;
+
+import com.bazaarvoice.emodb.common.uuid.TimeUUIDs;
+import com.bazaarvoice.emodb.kafka.KafkaCluster;
+import com.bazaarvoice.emodb.kafka.Topic;
+import com.bazaarvoice.megabus.guice.MegabusRefTopic;
+import com.bazaarvoice.megabus.resource.Coordinate;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.Futures;
+import com.google.inject.Inject;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Clock;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Default implementation which uses the existing Megabus actors.
+ */
+public class DefaultMegabusSource implements MegabusSource {
+
+    private static final Logger _LOG = LoggerFactory.getLogger(DefaultMegabusSource.class);
+
+    private final Producer<String, JsonNode> _producer;
+    private final Topic _topic;
+    private final ObjectMapper _objectMapper;
+    private final Clock _clock;
+
+    @Inject
+    public DefaultMegabusSource(KafkaCluster kafkaCluster, @MegabusRefTopic Topic topic,
+                                ObjectMapper objectMapper, Clock clock) {
+        _producer = requireNonNull(kafkaCluster.producer(), "producer");
+        _objectMapper = requireNonNull(objectMapper, "objectMapper");
+        _topic = requireNonNull(topic, "topic");
+        _clock = requireNonNull(clock, "clock");
+    }
+
+    /*
+     * Send the given co-ordinate to the Megabus Ref Topic.
+     */
+    @Override
+    public void touch(String table, String key) {
+        Coordinate coordinate = new Coordinate(table, key);
+        touchAll(Collections.singletonList(coordinate).iterator());
+    }
+
+    /*
+     * Send the given input refs to the Megabus Ref Topic.
+     */
+    @Override
+    public void touchAll(Iterator<Coordinate> coordinates) {
+        List<Coordinate> coordinateList = Lists.newArrayList();
+        coordinates.forEachRemaining(coordinateList::add);
+
+        _LOG.info("Sending {} coordinate(s) to Megabus Ref Topic: {}", coordinateList.size(), _topic.getName());
+        List<Future> futures = coordinateList.stream()
+                // Using the minimum UUID here to make sure the time is always beyond the FCL so that the resolver is certain to put the document in to actual Megabus.
+                // This way we wouldn't be in a situation where there is a ref in Ref topic but not in the Megabus topic.
+                .map(coordinate -> new MegabusRef(coordinate.getTable(), coordinate.getKey(), TimeUUIDs.minimumUuid(), _clock.instant()))
```

Comment:
This logic is pretty complicated, and it is not 100% right, we will corrupt the megabus. It's probably a good idea to limit this logic to a single function in a DAO-like singleton. We could great a single `MegabusRefProducerDAO` that just accepts `MegabusRef`'s and writes them to the appropriate place.